added options for filenames based on fiddle title

rudifa opened the pull request:
Hello Victor,

Having used your jsfiddle-downloader for a while, I decided to add options that let the user
specify additional filename forms, based on the fiddle title, which I find more informative.

The effects of the new options are described in --help.

There are no new dependencies for jsfiddle-downloader.js.
There are a couple of devDependencies (compare-version and fs-extra) for test.js.

I tested my changes only on macOS 10.13.2 High Sierra.

I also added test/test.js to verify the behavior of existing and of new options.

Please let me know whether you would consider my changes for inclusion in a future release
of jsfiddle-downloader. Your criticism and suggestions for improvements will be welcome
in any case.

Rudi Farkas


Here are some usage examples:    

```
jsfiddle-downloader-master $ bin/jsfiddle-downloader.js --help.


  Usage: jsfiddle-downloader [options]

  Options:

    -h, --help                        output usage information
    -V, --version                     output the version number
    -u, --user <user>                 Save all the users fiddles
    -l, --link <url>                  Url of the fiddle to save
    -o, --output <path>               Target path to download the data
    -c, --compressed                  Compress the spaces of the HTML output
    -i, --identifier <fiddle_id>      Identifier of the fiddle to save
    -f, --force-http                  Use http when the URI method is undefined
    -v, --verbose                     Verbose output
    -I, --filename-identifier         Use fiddle identifier as filename (default)
    -T, --filename-title              Use fiddle title as filename
    -IT, --filename-identifier-title  Use fiddle identifier and title as filename
    -S, --filename-spaces             Keep spaces in filename (default: replace by underscores)
```

Download a fiddle using the new options:

```
jsfiddle-downloader-master $ 
jsfiddle-downloader-master $ bin/jsfiddle-downloader.js --version
0.1.7
jsfiddle-downloader-master $ bin/jsfiddle-downloader.js -fi jghfjgcu
Output file = /Users/rudifarkas/Downloads/jsfiddle-downloader-master/jghfjgcu.html
jsfiddle-downloader-master $ bin/jsfiddle-downloader.js -fi jghfjgcu -IT
Output file = /Users/rudifarkas/Downloads/jsfiddle-downloader-master/jghfjgcu_Contadores_con_CSS_by_FacuCode.html
jsfiddle-downloader-master $ bin/jsfiddle-downloader.js -fi jghfjgcu -T
Output file = /Users/rudifarkas/Downloads/jsfiddle-downloader-master/Contadores_con_CSS_by_FacuCode.html
jsfiddle-downloader-master $ bin/jsfiddle-downloader.js -fi jghfjgcu -TS
Output file = /Users/rudifarkas/Downloads/jsfiddle-downloader-master/Contadores con CSS by FacuCode.html
jsfiddle-downloader-master $ bin/jsfiddle-downloader.js -fi jghfjgcu -ITS
Output file = /Users/rudifarkas/Downloads/jsfiddle-downloader-master/jghfjgcu_Contadores con CSS by FacuCode.html
```
Run tests:

```
jsfiddle-downloader-master $ mocha test


  jsfiddle-downloader v0.1.6 and above has features as described in README
  version detected 0.1.7

    ✓ 0.1.6-1 Download a single fiddle from its id, options -fi: (735ms)
    ✓ 0.1.6-2 Download a single fiddle from its id, options -fi -o: (724ms)
    ✓ 0.1.6-3 Download a single fiddle from its url, options -fl: (1491ms)
    ✓ 0.1.6-4 Download a single fiddle from its url, options -fl -o: (1505ms)
    ✓ 0.1.6-5 Download a single fiddle from its url, options -fl: (712ms)
    ✓ 0.1.6-6 Download a single fiddle from its url, options -fl: (775ms)
    ✓ 0.1.6-7 Download a single fiddle from its url, options -fl: (969ms)
    ✓ 0.1.6-8 Download a single fiddle from its url, options -fl: (652ms)
    ✓ 0.1.6-9 Download a single fiddle from its url, options -fl: (776ms)
    ✓ 0.1.6-10 Download all scripts of a user from jsFiddle.net, options -fu: (4411ms)
    ✓ 0.1.6-11 Download all scripts of a user from jsFiddle.net, options -fu -o: (2246ms)

  jsfiddle-downloader v0.1.7 has additional features as described in README
  version detected 0.1.7

    ✓ 0.1.7-1 Download a single fiddle from its id, options -fi -I: (666ms)
    ✓ 0.1.7-2 Download a single fiddle from its id, options -fi -T: (770ms)
    ✓ 0.1.7-3 Download a single fiddle from its id, options -fi -TS: (830ms)
    ✓ 0.1.7-4 Download a single fiddle from its id, options -fi -IT: (674ms)
    ✓ 0.1.7-5 Download a single fiddle from its url, options -fl -I: (761ms)
    ✓ 0.1.7-6 Download a single fiddle from its url, options -fl -T: (724ms)
    ✓ 0.1.7-7 Download a single fiddle from its url, options -fl -TS: (775ms)
    ✓ 0.1.7-8 Download a single fiddle from its url, options -fl -IT: (717ms)
    ✓ 0.1.7-9 Download all scripts of a user from jsFiddle.net, options -fu -I: (2353ms)
    ✓ 0.1.7-10 Download all scripts of a user from jsFiddle.net, options -fu -T: (2212ms)
    ✓ 0.1.7-11 Download all scripts of a user from jsFiddle.net, options -fu -TS: (2567ms)
    ✓ 0.1.7-12 Download all scripts of a user from jsFiddle.net, options -fu -IT: (2189ms)


  23 passing (31s)
```

rudifa committed:
added options for filenames based on fiddle title

facundovictor commented on package.json:
```diff
@@ -1,13 +1,13 @@
   },
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "mocha"
```
`mocha` is missing in the `devDependencies`.

facundovictor commented on bin/jsfiddle-downloader.js:
```diff
@@ -45,6 +49,27 @@ function printError(str){
+        }
+        return global.cwd+'/'+output+'.html'
+    }
+}
```
Nice! I like the idea of taking the title from the file. I'd need to test it with weird names to see how that goes. Maybe we could try getting read of uncommon characters.

facundovictor commented on bin/jsfiddle-downloader.js:
```diff
@@ -133,7 +158,7 @@ function getListOfFiddles(user){
             res.setEncoding('utf8');
             res.on('data', function (chunk) {
-                logIfVerbose('Retreive chunk');
+                logIfVerbose('Retrieve chunk');
```
#10 Good catch!

facundovictor commented on package-lock.json:
```diff
@@ -0,0 +1,870 @@
+  "version": "0.1.7",
+  "lockfileVersion": 1,
+  "requires": true,
+  "dependencies": {
```
This is a nice addition. But, we should use `npm-shrinkwrap`, as this is a CLI tool. See https://docs.npmjs.com/files/shrinkwrap.json. This would also help to keep backward compatibility with older NPM versions.

facundovictor commented on test/test.js:
```diff
@@ -0,0 +1,285 @@
+        if (error) {
+            // console.log(error.stack);
+            // console.log('Error Code: '+error.code);
+            // console.log('Reason of error: '+error.signal);
```
We should not leave commented code.

rudifa replied:
I agree.

facundovictor commented on test/test.js:
```diff
@@ -0,0 +1,285 @@
+          console.log(`  version detected ${stdout}`);
+          if (stderr != '') {
+              __this.skip();
+          } else if (compare_version(stdout, '0.1.6') < 0) {
```
Adding tests is one of the things that we must do if a tool is starting being used by more people.
But, the tests should be targeting features, not versions. Adding special conditions for each `patch` update would add unnecessary complexity. Anyway, I truly appreciate your effort, this definitely will push the app to the right direction after a couple of tweaks.

rudifa replied:
I agree with your principle "the tests should be targeting features, not versions".

I grouped tests by versions so that I could verify that (1) features present in '0.1.6' work in both '0.1.6' and '0.1.7', and (2) that features present only in '0.1.7' also work.

If you adopt and merge my changes, then all tests can be grouped together, after a simple test verifying that the version is at least '0.1.7'.

facundovictor commented on test/test.js:
```diff
@@ -0,0 +1,285 @@
+//#############################################################################
+// require helper node_modules and prepare helper methods and consts
+
+const expect = require('chai').expect;
```
`chai` seems to not be in the `package.json`.

facundovictor reviewed (commented):
Thanks @rudifa for contributing! I appreciate it!

I added some comments. But, It happened time since you made the PR, so I'll be fixing/merging it.

facundovictor committed:
Add chai and mocha to devDependencies

facundovictor committed:
Merge branch 'develop' into filename-options

facundovictor committed:
Add sanitize title for being used as filename

```sh
Control characters (0x00–0x1f and 0x80–0x9f)
Reserved characters (/, ?, <, >, \, :, *, |, and ")
Unix reserved filenames (. and ..)
Trailing periods and spaces (for Windows)
Windows reserved filenames (CON, PRN, AUX, NUL, COM1, COM2, COM3, COM4,
COM5, COM6, COM7, COM8, COM9, LPT1, LPT2, LPT3, LPT4, LPT5, LPT6, LPT7,
LPT8, and LPT9)
```

See: https://www.npmjs.com/package/sanitize-filename#details

facundovictor committed:
Add npm-shrinkwrap.json

We should use nppm-shrinkwrap instead of package-lock.json, as this
tool is a CLI tool that should guarantee all dependencies and nested
dependencies are preserved when users install this from npm.

facundovictor committed:
Remove commented code

facundovictor committed:
Remove package-lock.json in favor of npm-shrinkwrap
